microceph/ceph/osd.go

```diff
@@ -212,6 +212,46 @@ func checkEncryptSupport() error {
 	return nil
 }
 
+// setHostFailureDomain sets the host failure domain for the given host.
+func setHostFailureDomain() error {
+	var err error
+
+	if !haveCrushRule("microceph_auto_host") {
+		err = addCrushRule("microceph_auto_host", "host")
+		if err != nil {
+			return err
+		}
+	}
+	osdPools, err := getPoolsForDomain("osd")
+	if err != nil {
+		return err
+	}
+	for _, pool := range osdPools {
+		err = setPoolCrushRule(pool, "microceph_auto_host")
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// updateFailureDomain checks if we need to update the crush rules failure domain.
+// Once we have at least 3 nodes with at least 1 OSD each, we set the failure domain to host.
+func updateFailureDomain(s *state.State) error {
+	numNodes, err := database.MemberCounter.Count(s)
+	if err != nil {
+		return fmt.Errorf("Failed to count members: %w", err)
+	}
+
+	if numNodes >= 3 {
```

Comment:
I think scaling up is one thing that's easy to do in an automated sense, but scaling back is something different as you really run the risk of destroying data. As @lmlg indicates, it's out of the scope of this PR - but I think it should really be discussed and designed carefully such that the user is conscious of removing the OSDs may have impact to the cluster data.

Reply:
Indeed, scaling down needs to be handled separately. At the moment MicroCeph doesn't support scale-down; we created separate stories for this to be implemented safely. Part of that will be to handle the situation where we go below what host failure domain can handle i.e. 2 nodes or less. I'm partial to handle this symmetrically: handle  scale down as the reverse from scale up and go from failure domain=host to failure domain=osd. But this will need to be specced out in detail.